R/variables.R

```diff
@@ -0,0 +1,323 @@
+#' @export
+nvariables.NULL <- function(x, ...) {
+  0
+}
+
+#' @export
+nvariables.draws <- function(x, ...) {
+  length(variables(x, ...))
+}
+
+
+# internal ----------------------------------------------------------------
+
+# check validity of existing variable names: e.g., that
+# all `variables` exist in `x` and that no `variables`are reserved words
+# Additionally, this returns the cannonical name, so e.g. "theta" will get
+# converted to c("theta[1]", "theta[2]", ...) if those variables exist.
+# @param regex should 'variables' be treated as regular expressions?
+# @param scalar_only should only scalar variables be matched?
+check_existing_variables <- function(variables, x, regex = FALSE,
+                                     scalar_only = FALSE, exclude = FALSE) {
+  check_draws_object(x)
+  if (is.null(variables)) {
+    return(NULL)
+  }
+
+  regex <- as_one_logical(regex)
+  scalar_only <- as_one_logical(scalar_only)
+  exclude <- as_one_logical(exclude)
+  variables <- unique(as.character(variables))
+  all_variables <- variables(x, reserved = TRUE)
+
+  if (regex) {
+    tmp <- named_list(variables)
+    for (i in seq_along(variables)) {
+      tmp[[i]] <- grep(variables[i], all_variables)
+    }
+    # regular expressions are not required to match anything
+    missing_variables <- NULL
+    variables <- as.character(all_variables[unique(unlist(tmp))])
+  } else if (!scalar_only) {
+    # need to find variables that are matched by either a scalar or vector
+    # variable in x and what the matching variable is, while keeping original
+    # order of input `variables`
+
+    # find scalar variables (1-to-1 match between all_variables and variables)
+    scalar_input_ixs <- match(all_variables, variables)
+    # find vector variable matches (match all_variables with the indexing stripped)
+    all_variables_base <- all_variables
+    # exclude already matched scalar variables
+    all_variables_base[!is.na(scalar_input_ixs)] <- NA_character_
+    all_variables_base <- split_variable_names(all_variables_base)$base_name
+    vector_input_ixs <- match(all_variables_base, variables)
+    # compose the vector of indices of matched input variables
+    input_ixs <- c(scalar_input_ixs[!is.na(scalar_input_ixs)],
+                   vector_input_ixs[!is.na(vector_input_ixs)])
+    # compose the vector of indices of matched all_variables
+    all_var_ixs <- seq_along(all_variables)
+    all_var_matched_ixs <- c(all_var_ixs[!is.na(scalar_input_ixs)],
+                             all_var_ixs[!is.na(vector_input_ixs)])
+    # select missed input variables
+    missing_vars_mask <- rep_len(TRUE, length(variables))
+    missing_vars_mask[input_ixs] <- FALSE
+    missing_variables <- variables[missing_vars_mask]
+    # select matched all_variables maintaining the input variables order
+    variables <- all_variables[all_var_matched_ixs[order(input_ixs, all_var_matched_ixs)]]
+  } else {
+    missing_variables <- setdiff(variables, all_variables)
```

Comment:
This line is not hit by test coverage (see [here](https://app.codecov.io/gh/stan-dev/posterior/pull/342/blob/R/variables.R#L279)), but as far as I can tell it cannot be, as `scalar_only` is always `FALSE` (I assume this is a parameter that was used by some old code?). 

I have not changed this function except to use the new `split_variable_names()` on line 263. It only appears as all new because I moved it here from `R/draws-index.R`.

Reply:
I don't remember scalar_only. The way the variables are named suggests that I have I not written it. Was it previously used in rvars perhaps? I am fine with removing it as long as you don't see any value in keeping it for future use in rvars or similar functionality.

Reply:
I see. Interesting. I don't have any strong opinions about it. @jgabry and @MansMeg you seem to have been involved in the discussion. Any thoughts on this?